src/scripts/clipperUI/components/modeButtonSelector.tsx

```diff
@@ -31,101 +42,150 @@ class ModeButtonSelectorClass extends ComponentBase<{}, ClipperStateProp> {
-			myMode={ClipMode.Selection} tabIndex={43} selected={currentMode === ClipMode.Selection}
-			onModeSelected={this.onModeSelected.bind(this) }
-			tooltipText={Localization.getLocalizedString("WebClipper.ClipType.Selection.Button.Tooltip")}/>;
+		return {
+			imgSrc: ExtensionUtils.getImageResourceUrl("select.png"),
+			label: Localization.getLocalizedString("WebClipper.ClipType.Selection.Button"),
+			myMode: ClipMode.Selection,
+			selected: currentMode === ClipMode.Selection,
+			onModeSelected: this.onModeSelected.bind(this),
+			tooltipText: Localization.getLocalizedString("WebClipper.ClipType.Selection.Button.Tooltip")
+		};
 	}
 
-	private getBookmarkModeButton(currentMode: ClipMode) {
+	private getBookmarkButtonProps(currentMode: ClipMode): NoAriaPropsForModeButton {
 		if (this.props.clipperState.pageInfo.rawUrl.indexOf("file:///") === 0) {
 			return undefined;
 		}
-		return <ModeButton imgSrc={ExtensionUtils.getImageResourceUrl("bookmark.png") }
-			label={Localization.getLocalizedString("WebClipper.ClipType.Bookmark.Button") }
-			myMode={ClipMode.Bookmark} tabIndex={44} selected={currentMode === ClipMode.Bookmark}
-			onModeSelected={this.onModeSelected.bind(this) }
-			tooltipText={Localization.getLocalizedString("WebClipper.ClipType.Bookmark.Button.Tooltip") } />;
+
+		return {
+			imgSrc: ExtensionUtils.getImageResourceUrl("bookmark.png"),
+			label: Localization.getLocalizedString("WebClipper.ClipType.Bookmark.Button"),
+			myMode: ClipMode.Bookmark,
+			selected: currentMode === ClipMode.Bookmark,
+			onModeSelected: this.onModeSelected.bind(this),
+			tooltipText: Localization.getLocalizedString("WebClipper.ClipType.Bookmark.Button.Tooltip")
+		};
 	}
 
-	public render() {
+	private getListOfButtons() {
 		let currentMode = this.props.clipperState.currentMode.get();
 
+		let buttonProps = [
+			this.getFullPageButtonProps(currentMode),
+			this.getRegionButtonProps(currentMode),
+			this.getAugmentationButtonProps(currentMode),
+			this.getSelectionButtonProps(currentMode),
+			this.getBookmarkButtonProps(currentMode),
+			this.getPdfButtonProps(currentMode),
+		];
+
+		let visibleButtons = [
+			this.getScreenReaderOnlyElementPropsThatAnnouncesCurrentMode(currentMode),
+		];
+
+		if (buttonProps) {
+			let propsForVisibleButtons = buttonProps.filter(attributes => !!attributes);
+			for (let attributes of propsForVisibleButtons) {
+				let ariaPos = ModeButtonSelectorClass.getIndexOf(propsForVisibleButtons, attributes) + 1;
+				visibleButtons.push(<ModeButton imgSrc={attributes.imgSrc} label={attributes.label} myMode={attributes.myMode}
+								selected={attributes.selected} onModeSelected={attributes.onModeSelected}
+								tooltipText={attributes.tooltipText} aria-setsize={propsForVisibleButtons.length}
+								aria-posinset={ariaPos} tabIndex={attributes.selected ? 40 : undefined} />);
+			}
+			return visibleButtons;
+		}
+	}
+
+	private static getIndexOf(list, element): number {
```

Comment:
This is still O(1). Rather than using for ... of in the above block, using a simple for loop will remove the need to iterate through the list again to find the index.